File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-api/src/main/java/org/kie/workbench/common/services/refactoring/service/ResourceType.java
/*
 * Copyright 2016 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kie.workbench.common.services.refactoring.service;

import org.jboss.errai.common.client.api.annotations.Portable;

@Portable
public enum ResourceType {

    /**
     * jBPM
     */
    BPMN2("bpmn2id"),
    // the problem is that processes can be referred to
    // 1. by id
    // OR
    // 2. by name
    // because of that, we add both when indexing
    BPMN2_NAME("bpmn2name"),

    /**
     * Drools
     */
    RULE("rule"),
    FUNCTION("function"),
    QUERY("query"),
    DRL_ENUM("drlenum"),
    DRL_ANNOTATION("drlanno"),

    /**
     * Form (as in, form-modeller)
     */
    FORM("form"),

    /**
     * Java file
     */
    JAVA("java"),

    /**
     * Properites configuration files
     */
    PROPERTIES_OR_CONFIG("properties"),

    /**
     * Case Modeler
     */
    BPMN_CM("bpmncmid"),

    BPMN_CM_NAME("bpmncmname");


    /**
     * All luceneValue values (used in indexing) should be lowercase
     */
    private final String luceneValue;

    /**
     *
     */
    private ResourceType(String value) {
        this.luceneValue = value;
    }

    @Override
    public String toString() {
        return this.luceneValue;
    }

}

File: controllers/workflowdef/resource_type.go
// Copyright 2023 Red Hat, Inc. and/or its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package workflowdef

import (
	"strings"

	"github.com/kiegroup/kogito-serverless-operator/api/metadata"
)

type ExternalResourceType string

const (
	ExternalResourceCamel    ExternalResourceType = "resource-camel"
	ExternalResourceOpenApi  ExternalResourceType = "resource-openapi"
	ExternalResourceAsyncApi ExternalResourceType = "resource-asyncapi"
	ExternalResourceGeneric  ExternalResourceType = "resource-generic"
	ExternalResourceNone     ExternalResourceType = ""
)

// ExternalResourceCamelMountDir directory where to mount Camel Routes resource files
const ExternalResourceCamelMountDir = "routes"

// ExternalResourceDestinationDir map for special directories within the resource context.
// In dev mode means within the src/main/resources. In build contexts, the actual context dir.
var ExternalResourceDestinationDir = map[ExternalResourceType]string{
	ExternalResourceGeneric:  "",
	ExternalResourceCamel:    ExternalResourceCamelMountDir,
	ExternalResourceOpenApi:  "",
	ExternalResourceAsyncApi: "",
}

const externalResourceAnnotationPrefix = "resource"

// GetAnnotationResourceType gets the resource type for the given annotation.
// Example: Workflows can be annotated with sw.kogito.kie.org/resource-openapi=MyOpenApisConfigMapName.
// This means that a ConfigMap named MyOpenApisConfigMapName is available in the current context, and it will be mounted in the workflow.
// This function returns "resource-openapi" for the given example above.
func GetAnnotationResourceType(annotationKey string) ExternalResourceType {
	stringArray := strings.Split(annotationKey, "/")
	if len(stringArray) == 2 && stringArray[0] == metadata.Domain && strings.HasPrefix(stringArray[1], externalResourceAnnotationPrefix) {
		return ParseExternalResourceType(stringArray[1])
	} else {
		return ""
	}
}

// GetExternalResourceTypeAnnotation gets the key for the ExternalResourceType annotation.
func GetExternalResourceTypeAnnotation(resourceType ExternalResourceType) string {
	return metadata.Domain + "/" + string(resourceType)
}

// ParseExternalResourceType ...
func ParseExternalResourceType(resourceType string) ExternalResourceType {
	switch resourceType {
	case string(ExternalResourceCamel):
		return ExternalResourceCamel
	case string(ExternalResourceOpenApi):
		return ExternalResourceOpenApi
	case string(ExternalResourceAsyncApi):
		return ExternalResourceAsyncApi
	case string(ExternalResourceGeneric):
	default:
		return ExternalResourceNone
	}
	return ExternalResourceNone
}
